R/utils-co2_upper_lower.R

```diff
@@ -0,0 +1,13 @@
+set.seed(111)
```

Comment:
@maurolepore I want to get reproducible results from the function `jitter_range()`.  This function always create random values after each new run. How do you propose that I should put the seed for this function? 

You can find the function [here](https://github.com/2DegreesInvesting/tiltIndicator/blob/5ae818397fc64f9cb9e6b5ad3540045025658843/R/jitter_range.R#L26). Although, Thanks for your suggestion on applying seed in tests.

Reply:
The name `jitter` implies the function produces a random result. But if you set the seed inside the function then the function no longer produces a random result to the name becomes misleading.

``` r
# This function is rigid
f <- function() {
  set.seed(1)
  
  sample(1:10)
}

# This function is no longer random
f()
#>  [1]  9  4  7  1  2  5  3 10  6  8
f()
#>  [1]  9  4  7  1  2  5  3 10  6  8

# This limits the ability of the users to change the seed themselves
set.seed(2)
# Still the same
f()
#>  [1]  9  4  7  1  2  5  3 10  6  8



# This new implementation gives back the control of randomness to the users
g <- function() {
  sample(1:10)
}

# If they want the same result they set the same seed
set.seed(2)
g()
#>  [1]  5  6  9  1 10  7  4  8  3  2
set.seed(2)
g()
#>  [1]  5  6  9  1 10  7  4  8  3  2

# If they want random results they don't set the seed
g()
#>  [1]  1  3  6  2  9 10  7  5  4  8
g()
#>  [1]  6  9 10  5  8  3  1  7  2  4
```

--

The time to set the seed is at the time of using the function, not at the time of defining it. So you wouldn't use it when you wear your developer hat. you would use it when you wear your analyst hat -- e.g. when you write a script to process real data you would use `set.seed()` at the top of that script.

As a developer you should only need it in tests, for example if you take a snapshot of some randomly-generated data.

Reply:
@maurolepore Thanks for this insight! Its very clear to me now. Your explanation means I used the `set.seed` at the right place on top of the script. However, I will write a test to ensure that the results are reproducible on the sample data. Does it sounds good?

Reply:
@maurolepore As we create the final outputs using `tiltWorkflows`, should we set the seed then in `tiltWorkflow` package? We only use that analyst script in `tiltWorkflows` and not in `tiltIndicatorAfter`. And a test to ascertain reproducible results then can also only be written in tiltWorkflow? right?

Reply:
The same goes for tiltWorkflows. You may use `set.seed()` when you use the tiltWorkflows package to produce real output, but not when we write the tiltWorkflows package itself.

You may want to see how other people use (or not) `set.seed()` in their packages by searching for "seed" in a repo. For example:

- https://github.com/search?q=repo%3A2DegreesInvesting%2FtiltIndicator+seed&type=code
- https://github.com/search?q=repo%3Atidyverse%2Fdplyr%20seed&type=code

Notice that developers use the seed mostly in files that simulate the experience of a user -- like examples, vignettes, and tests -- but not in code that belongs to the package (i.e. not under the R/ directory).

Reply:
Maybe it would help to make concrete your need to set the seed. I suggest you go ahead and push a test (in this PR) that fails because you get inconsistent results because of the randomly-generated data. Then we'll add `withr::local_seed()` in that specific test and it should pass.

Reply:
@maurolepore Thanks for your insights through the call as well! I have added the tests which ensure that we get reproducible results after setting the seed by the user. 

However, @AnneSchoenauer and @Tilmon I need your insight on this one remaining issue:
Do we want the outputs in `co2e_lower` and `co2e_upper` to be random and reproducible? or in other words, Do we want the user to see the same results in these columns which are randomised?

Reply:
Hi all, @kalashsinghal @maurolepore @AnneSchoenauer , 

Kalash and I had a quick call and discussed that the outputs in `co2_lower `and `co2_upper ` can be random for now but should be made reproducible in the future. This however is anyways a separate task as the set.seed() function would need to be included in tiltWorkflows. 

We'll prioritize that task in the next sprint.